### PR TITLE
fix(core): Evenly distribute delete resources across the configured delete spread time divided by the total number of resource partitions

### PR DESCRIPTION
**Pull request by jonsie**
Previously, we had jitter across a fixed range that was added to these deletes.  This still ended up with deletes being clumped into a fairly tight window, which caused AWS rate limiting in downstream services.  This proposes to delete across an evenly distributed time window.

### CODE REVIEW

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/AbstractResourceTypeHandler.kt [hunk 1]
```diff
@@ -527,21 +528,33 @@ abstract class AbstractResourceTypeHandler<T : Resource>(
...
+        partitions.add(ResourcePartition(markedResources = groupedMarkedResources))
       }
     }
 
-    return partitions.sortedByDescending { it.size }
+    if (partitions.isNotEmpty()) {
+      val offsetIntervalMs = configuration.getDeleteSpreadMs(dynamicConfigService) / partitions.size
+      var offsetMs = 0L
+      partitions.forEach { partition ->
+        if (offsetMs != 0L) offsetMs += offsetIntervalMs
```
Comment (dreynaud) [inline swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/AbstractResourceTypeHandler.kt:552]: this conditional seems odd, we want to add the `offsetIntervalMs` on every iteration, right?

Comment (jonsie) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/AbstractResourceTypeHandler.kt:552]: Oh wait I thought I deleted that. One moment.

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/AbstractResourceTypeHandler.kt [hunk 2]
```diff
@@ -527,21 +528,33 @@ abstract class AbstractResourceTypeHandler<T : Resource>(
...
+        partition.offsetMs = offsetMs
+      }
+    }
```
Comment (dreynaud) [inline swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/AbstractResourceTypeHandler.kt:555]: How about this instead?
```suggestion
      partitions.forEach { partition ->
        partition.offsetMs = offsetMs
        offsetMs += offsetIntervalMs
      }
    }
```

Comment (jonsie) [reply swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/AbstractResourceTypeHandler.kt:555]: Yes yes.

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/AbstractResourceTypeHandler.kt [hunk 3]
```diff
@@ -527,21 +528,33 @@ abstract class AbstractResourceTypeHandler<T : Resource>(
...
+      Lists.partition(it.value, configuration.itemsProcessedBatchSize).forEach { groupedMarkedResources ->
+        partitions.add(ResourcePartition(markedResources = groupedMarkedResources))
       }
     }
 
-    return partitions.sortedByDescending { it.size }
+    if (partitions.isNotEmpty()) {
+      val offsetIntervalMs = configuration.getDeleteSpreadMs(dynamicConfigService) / partitions.size
+      var offsetMs = 0L
+      partitions.forEach { partition ->
```
Comment (dreynaud) [inline swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/AbstractResourceTypeHandler.kt:551]: for some extra fancy pantsiness:

```suggestion
      partitions.forEachIndexed { index, partition ->
        partition.offsetMs = offsetMs + index * offsetIntervalMs
      }
```

**Comment by aravindmd** [comment]
Overall approach sounds good to me , might need a couple of deletion tests to be updated.

**Review by aravindmd — APPROVED** [review]

**Pull request merged**